src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs

```diff
@@ -305,6 +305,20 @@ private static void OnRequestError(WinHttpRequestState state, Interop.WinHttp.WI
             
             Debug.Assert(state != null, "OnRequestError: state is null");
 
+            Debug.Assert((unchecked((int)asyncResult.dwError) != Interop.WinHttp.ERROR_INSUFFICIENT_BUFFER &&
+                unchecked((int)asyncResult.dwError) != unchecked((int)0x80090321)), // SEC_E_BUFFER_TOO_SMALL
+                $"Unexpected async error in WinHttpRequestCallback: {unchecked((int)asyncResult.dwError)}, WinHttp API: {unchecked((uint)asyncResult.dwResult.ToInt32())}");
+            
+            // If the failure is not caused by the above error codes, need to get more information from asyncResult.dwResult.
+            // There are only two cases in the below switch statement which explicitly set WinHttpException.
+            Debug.Assert((unchecked((uint)asyncResult.dwResult.ToInt32()) != Interop.WinHttp.API_SEND_REQUEST &&
```

Comment:
PTAL here, the logic here is complicated. Previously it's like:
if ( (m == n) || ( ((x == y) && (a !=b) && (a != c) && a != d) ) ) {fail}

Now it's:
Debug.Assert( (m != n) && ((x != y) || (a == b) || (a == c) || (a == d)), some message )